### PR TITLE
Code review

### PR DESCRIPTION
**Pull request by sisiranair**
@ChristopherEeles Please review the code - good practices, additional info etc.

### CODE REVIEW

File: ClinicalTrial_SE_curation.Rmd [hunk 1]
```diff
@@ -0,0 +1,320 @@
+---
+title: "Curation of Clinical trial Sumamrized Experiment objects"
+author: "Nikta Feizi, Christopher Eeles, Sisira Nair"
+date: "`r format(Sys.time(), '%a/%d/%b')`"
+output: html_document
+---
+#Please note:
+Parameters of "run_all" function are data specific and must be defined by looking at the metadata outside of the function.
```
Comment (ChristopherEeles) [inline ClinicalTrial_SE_curation.Rmd]: Best Practices:

R functions should pass all necessary data as a parameter. You should never expect objects outside of the function definition to exist in the environment. This is opaque and makes it difficult to understand what your function is doing.

Instead, pass each piece of data in as a parameter. Ideally, you will write brief documentation in the form of comments for these parameters and update them whenever the function is changed. 

Note: `#'` specific a special kind of comment in R, a Roxygen2 comment. These can be parsed when building an R package to automatically produce documentation. It is a good idea to write your comments in this format. More information is available here: https://cran.r-project.org/web/packages/roxygen2/vignettes/roxygen2.html.

As an example of a minimum for a function:
```{r}
#' @param data1 file path to read data from
#' @param data2 file path to write data to
my_function <- function(data1, data2) { ... }
```

If the function returns something, it is a good idea to document this as well. If the function doesn't return anything, but instead works via a side effect such as writing to disk, this should also be documented.
```
#' @param data1 file path to read data from
#' @param data2 file path to write data to
#'
#' @return None, writes to file specific in data2
my_function <- function(data1, data2) { ... }
```

File: ClinicalTrial_SE_curation.Rmd [hunk 2]
```diff
@@ -0,0 +1,320 @@
...
+options(encoding = "UTF-8" , timeout = 1000)
+work_dir = setwd(paste0(getwd()))
+```
+
+```{r}
+# fetch Gencode v33 annotations from BHKLAB-Pachyderm/Annotations
+gencode_url <- "https://github.com/BHKLAB-Pachyderm/Annotations/raw/master/Gencode.v33.annotation.RData"
+gencode_file <- file.path(work_dir, "Ensembl.v99.annotation.RData")
+download.file(gencode_url, destfile=gencode_file)
+load("Ensembl.v99.annotation.RData")
```
Comment (ChristopherEeles) [inline ClinicalTrial_SE_curation.Rmd]: Suggestion:

TL;DR - Always use `load("path_to_file.RData", verbose=TRUE)`

Loading data from .RData files can be a bit confusing because there is no guarantee that the loaded objects have names related to the name of the file. This is why it is almost always preferable to save data to an `.rds` file, where you assign it to a variable during loading.

I obviously don't expect you to change the data object for this review, but a good work around to make sure it is clear how all objects in your script were instantiated is to set the `verbose=TRUE` flag when calling the `load` function. This will print the names of the loaded objects, making it clear where those object originated.

File: ClinicalTrial_SE_curation.Rmd [hunk 3]
```diff
@@ -0,0 +1,320 @@
...
+library(SummarizedExperiment)
+library(qs)
+library(affy)
+```
+
+```{r}
+geo_rma<- function (raw_tar, geo_path , brain_path){
+  
+  # Download CEL files
+  cel.file <- paste(work_dir, raw_tar, sep="/")
```
Comment (ChristopherEeles) [inline ClinicalTrial_SE_curation.Rmd]: Best Practices:

TL;DR - When modifying global options inside a function, ALWAYS add this before changing them with `options`
```
opts <- options()
on.exit(options(opts))
```

Setting options using the `options` function modifies a users global environment. Because functions in R should be totally self contained, and NOT change the global state of a users R environment after running, it is necessary to reset them when the function execution ends.

This can be easily achieved with the `on.exit` function, which registers an R expression to be run when the current function finishes executing.

First, you need to capture the current user options before modifying them:
`opts <- options()`

Then you register an expression to reset the user options when exiting a function:
`on.exit(options(opts))`

Then you can modify whichever options you need to for your function to work correctly.
```options( ... )```

Bugs caused when functions modify the global state of an R session are some of the hardest to figure out, and can end up wasting HOURS of debugging time.

Comment (ChristopherEeles) [inline ClinicalTrial_SE_curation.Rmd]: Best Practices:

Always build file paths with `file.path` because it does not assume a file separator for your OS. Linux and macOS use "/", but Windows uses "\\". `file.path` will read the file separator for the OS and is thus operating system agnostic.

File: ClinicalTrial_SE_curation.Rmd [hunk 4]
```diff
@@ -0,0 +1,320 @@
...
+  untar(raw_tar, exdir=exit_dir)
+  cel.files<-list.files(paste(exit_dir, "/", sep=""), pattern = "gz")
+  sapply(paste(exit_dir, cel.files, sep="/"), GEOquery::gunzip)
+  
+  # Install the cdf brainarray package
+  pack_name <- tail(str_split(brain_path, "/")[[1]], n=1) 
+  cdf.file <- paste(work_dir, pack_name, sep="/")
+  options(download.file.method="curl", download.file.extra="-k -L")
+  download.file( brain_path, cdf.file)
+  install.packages(pack_name, repos = NULL, type = "source")
```
Comment (ChristopherEeles) [inline ClinicalTrial_SE_curation.Rmd:79]: Best Practices:

Installing packages modifies a users global state. Please always make a comment above the function definition altering a user to this.

The best way is to add a `@return` tag like:
```
#' @return None, installs packages to the global environment
installs_some_packages <- function() { ... }
```

File: ClinicalTrial_SE_curation.Rmd [hunk 5]
```diff
@@ -0,0 +1,320 @@
...
+  
+  # Ordering columns based on priority
+  pheno_data <- pheno_data[, c("unique_patient_ID", prior, colnames(pheno_data)[!colnames(pheno_data) %in% prior])]
+  
+  # Replacing "NA" with NA_character_ entries
+  pheno_data[pheno_data=="NA"]=NA
+  
+  
+  # Feature data
+  feat_data <- merge(features_gene, data.frame(gene_id = rownames(assay_data)), all.y=T, by = "gene_id")
```
Comment (ChristopherEeles) [inline ClinicalTrial_SE_curation.Rmd]: Best Practices:

Never use `T` or `F` in an R script or function, they are only for interactive use. Take the time to write out `TRUE` or `FALSE` to enhance the clarity of your code.

File: ClinicalTrial_SE_curation.Rmd [hunk 6]
```diff
@@ -0,0 +1,320 @@
...
+  stopifnot(all(rownames(rowData(SE)) == rownames(fData(eSet))))
+  stopifnot(all(rownames(colData(SE)) == colnames(assay(SE))))
+  
+  return(SE)
+}
+```
+
+```{r}
+# Run all the functions
+run_all <- function(raw_tar, geo_path, brain_path, delim, GPL, cancer_type ,treatment, response, survival_type,
```
Comment (ChristopherEeles) [inline ClinicalTrial_SE_curation.Rmd]: Best Practices:

Always give your functions a descriptive name, it enhances readability and ensure you don't have to find the function definition to get an idea of what a function does.

This comment could apply to many functions in this file.

File: ClinicalTrial_SE_curation.Rmd [hunk 7]
```diff
@@ -0,0 +1,320 @@
...
+  SE <- eSetToSE(eset)
+  
+  return(SE)
+}
+
+```
+
+```{r}
+#==== 1 ====
+GSE41998_SE <- run_all(raw_tar = "GSE41998_RAW.tar" ,
```
Comment (ChristopherEeles) [inline ClinicalTrial_SE_curation.Rmd]: Suggestion:

It seems like many of the parameters you specified are the default. It is okay to exclude parameters from a function call if they are using the defaults for the function definition. This saves you time typing, and can also help focus on the relevant things being changed, enhancing the clarify of code for the reader.

File: ClinicalTrial_SE_curation.Rmd [hunk 8]
```diff
@@ -0,0 +1,320 @@
...
+```{r}
+#==== 5 ====
+GSE20194_SE <- run_all(raw_tar = "GSE20194_RAW.tar" , 
+                       geo_path = "https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSE20194&format=file" , 
+                       brain_path = "http://mbni.org/customcdf/24.0.0/ensg.download/hgu133ahsensgcdf_24.0.0.tar.gz",
+                       delim = "_.*" , GPL = "GPL96",cancer_type=NA, treatment = "treatment code:ch1" , response = "pcr_vs_rd:ch1",
+                       survival_type = NA, survival_time = NA, survival_time_unit= NA,  event_occured = NA)
+
+colData(GSE20194_SE)[["cancer_type"]] <- "Breast cancer"
+colData(GSE20194_SE)$treatment[is.na(colData(GSE20194_SE)$treatment) & colData(GSE20194_SE)$treatments.comments.ch1 == "Taxol x 12 FAC x 4"] <- "TFAC"
```
Comment (ChristopherEeles) [inline ClinicalTrial_SE_curation.Rmd:263]: Suggestions:

This not pleasant to read. It is often better to extract relevant data into variables with short but descriptive names to both save on typing and make code easier to read.

For example:
```
# extract the data
cData <- colData(GSE20194_SE)
has_na_treatment <- is.na(cData$treatment)
tx_comments <- cData$treatment.comments.ch1

# make the changes
cData[has_na_treatment & tx_comment == "Taxol x 12 FEC x4", ]$treatment <- "TFEC"
cData[has_na_treatment & tx_comment == "Taxol x 12  FEC x4", ]$treatment <- "TFEC"
cData[has_na_treatment & tx_comment == "FEC x 3 ( d/t N.C)Taxol x 12", ]$treatment <- "FECT"
cData[is.na(cData$treatment)]$treatment <- cData[is.na(cData$treatment)]$treatment.comments.chr

# update the object
colData(GSE20194_SE) <- cData
```

This also has the advantage that the actual SE colData is not modified until all the updates are complete, allowing you to check things are as you expect before modifying the actual SE.

NOTE: I have not tested this code, it is just an example of one way to improve readability.

File: ClinicalTrial_SE_curation.Rmd [hunk 9]
```diff
@@ -0,0 +1,320 @@
...
+                       delim = ".CEL" , GPL = "GPL8414", cancer_type=NA,treatment = "treatment:ch1", response = "response:ch1" , 
+                       survival_type = NA, survival_time = NA,  
+                       survival_time_unit= NA, event_occured = NA)
+
+colData(GSE15622_SE)[["cancer_type"]] <- "Ovarian Cancer"
+```
+
+
+```{r}
+#===== Save objects =======
```
Comment (ChristopherEeles) [inline ClinicalTrial_SE_curation.Rmd]: Suggestion:

This should probably be a for loop:
```
gse_datasets <- c("GSE41998_SE", "GSE14671_SE", ...)
for (dataset in gse_datasets) qsave(get(dataset), file=paste0(dataset, ".qs")
```

Also, `qsave` really isn't any better than `saveRDS` unless you set the `nthread` parameter like:
```
nthread <- 2
qsave(..., nthread=nthread)
```
This is also true for `qread`.

You can set the nthread as a variable at the top of your script to make it easy to change the threading everywhere.

File: ClinicalTrial_SE_curation.Rmd (hunk 1 above)
Comment (ChristopherEeles) [reply ClinicalTrial_SE_curation.Rmd]: It seems like maybe this has already been fixed. I didn't seen any functions with parameters that aren't defined. If this is the case, make sure to also update your comments.

The only thing worse than no comments is outdated comments.

**Review by ChristopherEeles — CHANGES_REQUESTED** [review]
Suggestions are optional.

Best practices are mandatory, but if we are short on time I will approve the merge as long as you take any unresolved best practice notes and turn them into issues on this repo.

File: ClinicalTrial_SE_curation.Rmd [hunk 10]
```diff
@@ -318,3 +318,17 @@ qsave(GSE25066_SE, "GSE25066_SE.qs")
 qsave(GSE15622_SE, "GSE15622_SE.qs")
 
 ```
+
+```{r}
+#===== Automatically upload SE objects to Zenodo =======
+remotes::install_github("bhklab/AnnotationGx@development")
```
Comment (ChristopherEeles) [inline ClinicalTrial_SE_curation.Rmd]: Suggestions:

To prevent reinstalling the package every time this script is run, maybe we want to use `require` like this:
```
if (!require(AnnotationGx)) remotes::install_github("bhklab/AnnotationGx@development")
```

File: ClinicalTrial_SE_curation.Rmd (hunk 1 above)
Comment (sisiranair) [reply ClinicalTrial_SE_curation.Rmd]: @NiktaFeizi

File: ClinicalTrial_SE_curation.Rmd (hunk 2 above)
Comment (sisiranair) [reply ClinicalTrial_SE_curation.Rmd]: The RData file - Ensembl.v99.annotation.RData is from https://github.com/BHKLAB-Pachyderm/Annotations and we might have to change it to rds. I have created an issue in that repo now.
For now, adding `verbose = TRUE`.

File: ClinicalTrial_SE_curation.Rmd (hunk 3 above)
Comment (sisiranair) [reply ClinicalTrial_SE_curation.Rmd]: @NiktaFeizi

Comment (sisiranair) [reply ClinicalTrial_SE_curation.Rmd]: @NiktaFeizi

File: ClinicalTrial_SE_curation.Rmd (hunk 4 above)
Comment (sisiranair) [reply ClinicalTrial_SE_curation.Rmd:79]: Done

File: ClinicalTrial_SE_curation.Rmd (hunk 5 above)
Comment (sisiranair) [reply ClinicalTrial_SE_curation.Rmd]: Done

File: ClinicalTrial_SE_curation.Rmd (hunk 6 above)
Comment (sisiranair) [reply ClinicalTrial_SE_curation.Rmd]: Done

File: ClinicalTrial_SE_curation.Rmd (hunk 7 above)
Comment (sisiranair) [reply ClinicalTrial_SE_curation.Rmd]: @NiktaFeizi

File: ClinicalTrial_SE_curation.Rmd (hunk 8 above)
Comment (sisiranair) [reply ClinicalTrial_SE_curation.Rmd:263]: @NiktaFeizi

File: ClinicalTrial_SE_curation.Rmd (hunk 9 above)
Comment (sisiranair) [reply ClinicalTrial_SE_curation.Rmd]: @NiktaFeizi

File: ClinicalTrial_SE_curation.Rmd (hunk 10 above)
Comment (sisiranair) [reply ClinicalTrial_SE_curation.Rmd]: Done

**Review by ChristopherEeles — DISMISSED** [review]
Not all comments need to be roxygen2 comments (`#'`), only function or object documentation.

File: ClinicalTrial_SE_curation.Rmd (hunk 3 above)
Comment (NiktaFeizi) [reply ClinicalTrial_SE_curation.Rmd]: Done

File: ClinicalTrial_SE_curation.Rmd (hunk 9 above)
Comment (NiktaFeizi) [reply ClinicalTrial_SE_curation.Rmd]: Done

File: ClinicalTrial_SE_curation.Rmd (hunk 3 above)
Comment (NiktaFeizi) [reply ClinicalTrial_SE_curation.Rmd]: Done

Comment (ChristopherEeles) [reply ClinicalTrial_SE_curation.Rmd]: I don't see these changes? Did you maybe forget to push them to the code_review branch?

File: ClinicalTrial_SE_curation.Rmd (hunk 1 above)
Comment (NiktaFeizi) [reply ClinicalTrial_SE_curation.Rmd]: Done

File: ClinicalTrial_SE_curation.Rmd (hunk 3 above)
Comment (ChristopherEeles) [reply ClinicalTrial_SE_curation.Rmd]: This was not quite correct. You need to capture the options inside the function before setting on.exit. Otherwise you have a function that requires a variable defined outside of its scope.

I have corrected it so we can get this PR completed.

**Review by ChristopherEeles — COMMENTED** [review]
```
opts <- options()
on.exit(options(opts))
```
Both lines need to live inside the function definition. I have corrected this already.

**Review by ChristopherEeles — APPROVED** [review]
Fixed some remaining issues.
